src/renderer/api/kube-watch-api.ts

```diff
@@ -43,50 +42,59 @@ export interface IKubeWatchLog {
 
 @autobind()
 export class KubeWatchApi {
-  private cluster: Cluster;
-  private namespaceStore: NamespaceStore;
-
   private requestId = 0;
-  private isConnected = false;
   private reader: ReadableStreamReader<string>;
-  private subscribers = observable.map<KubeApi, number>();
-
-  // events
   public onMessage = new EventEmitter<[IKubeWatchMessage]>();
 
+  @observable.ref private getCluster: () => Cluster;
+  @observable.ref private getNamespaces: () => string[];
+  @observable isConnected = false;
+  @observable subscribers = observable.map<KubeApi, number>();
+
+  @computed get isReady(): boolean {
+    return !!(this.cluster && this.namespaces);
+  }
+
+  @computed get cluster(): Cluster | undefined {
+    return this.getCluster?.();
+  }
+
+  @computed get namespaces(): string[] {
+    return this.getNamespaces?.() || [];
+  }
+
   @computed get isActive(): boolean {
     return this.apis.length > 0;
   }
 
   @computed get apis(): string[] {
-    const { cluster, namespaceStore } = this;
-    const activeApis = Array.from(this.subscribers.keys());
-
-    return activeApis.map(api => {
-      if (!cluster.isAllowedResource(api.kind)) {
+    return Array.from(this.subscribers.keys()).map(api => {
+      if (!this.isReady || !this.isAllowedApi(api)) {
```

Comment:
the `!this.isReady` check should probably be hoisted of the map function.

Reply:
Why?

Reply:
done